ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncDownloader.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.fastsync;
+
+import static tech.pegasys.pantheon.ethereum.eth.sync.fastsync.FastSyncError.FAST_SYNC_UNAVAILABLE;
+import static tech.pegasys.pantheon.ethereum.eth.sync.fastsync.FastSyncError.UNEXPECTED_ERROR;
+
+import tech.pegasys.pantheon.util.ExceptionUtils;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FastSyncDownloader<C> {
+  private static final Logger LOG = LogManager.getLogger();
+  private final FastSyncActions<C> fastSyncActions;
+
+  public FastSyncDownloader(final FastSyncActions<C> fastSyncActions) {
+    this.fastSyncActions = fastSyncActions;
+  }
+
+  public CompletableFuture<Optional<FastSyncError>> start() {
+    LOG.info("Fast sync enabled");
+    return fastSyncActions
+        .waitForSuitablePeers()
+        .thenApply(state -> fastSyncActions.selectPivotBlock())
+        .thenCompose(fastSyncActions::downloadPivotBlockHeader)
```

Comment:
I'm thinking about how the full sync downloader works and how that maps over to this downloader.  I think for the most part the flow should be the same.  In the full sync downloader, the first thing we do is find the "sync target".  We use this sync target to make sure that we're downloading a coherent chain (downloading the chain based off of hashes from our best peer rather than block numbers).  I think we probably want to download the pivot block based on a header that we get from our target peer.  We can still "confirm" that block header, but the confirmation would be based on the header / hash rather than block number.

Reply:
The risk is that we'll pick a header that's on a non-canonical chain.  If we request it by hash, other clients will likely have it and return it "confirming" it, but what we actually want is to check that they consider the canonical block at that number to be the same.

Reply:
Bear in mind that when you fast-sync to a block, you can't apply any re-org that goes back before the fast-sync point.  This was a problem in the Ropsten fork where geth clients wound up fast-syncing off Parity nodes and so got stuck on the incorrect chain with no ability to re-org back even when the correct chain eventually had a greater total difficulty.

Reply:
That's a good point!  Definitely makes sense to query for the pivot based on block number.  I guess my main point is just that whatever sync target we pick needs to agree that the pivot we find is on the main chain.   I was looking at this from the perspective of sync target, but just as valid to choose the pivot first and make sure our sync target is consistent with that.